### PR TITLE
[SYCL][FPGA] Add Intel FPGA bank_bits attribute

### PR DESCRIPTION
**Pull request by vmaksimo**
This attribute forces the memory system to implement the variable or struct
member in a memory with 2^(N+1) banks, with (b0, ..., bn) forming the
bank-select bits.

### CODE REVIEW

File: clang/include/clang/Basic/Attr.td [hunk 1]
```diff
@@ -1744,6 +1744,24 @@ def SYCLFPGAPipe : TypeAttr {
...
+  let Subjects = SubjectList<[IntelFPGAConstVar, IntelFPGALocalStaticSlaveMemVar,
+                              Field], ErrorDiag>;
+  let LangOpts = [SYCLIsDevice, SYCLIsHost];
+  let Documentation = [IntelFPGABankBitsDocs];
+  let AdditionalMembers = [{
+    static unsigned getMinValue() {
+      return 0;
+    }
+    static unsigned getMaxValue() {
+      return 1024*1024;
```
Comment (AGindinson) [inline clang/include/clang/Basic/Attr.td]: IMO, this limit should be documented. The current documentation doesn't imply this value, or, for that matter, any particular number of bits.

File: clang/test/SemaSYCL/intel-fpga-local.cpp [hunk 2]
```diff
@@ -372,6 +410,57 @@ void foo1()
...
+
+  //expected-error@+1{{bank_bits must be consecutive}}
+  [[intelfpga::bank_bits(1,3,4), intelfpga::bankwidth(4)]]
+  unsigned int bb_six[4];
+
+  //expected-error@+1{{attribute takes at least 1 argument}}
+  [[intelfpga::bank_bits]]
+  unsigned int bb_seven[4];
+
+  //expected-error@+1{{attribute takes at least 1 argument}}
```
Comment (MrSidims) [inline clang/test/SemaSYCL/intel-fpga-local.cpp]: I think it's redundant

File: clang/include/clang/Basic/Attr.td (hunk 1 above)
Comment (vmaksimo) [reply clang/include/clang/Basic/Attr.td]: As I see, the limits for another attributes were not documented, so I just followed the common documenting style :)

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 3]
```diff
@@ -5209,6 +5220,92 @@ static void handleIntelFPGAMergeAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
...
+  checkForDuplicateAttribute<IntelFPGABankBitsAttr>(S, D, Attr);
+
+  if (!checkAttributeAtLeastNumArgs(S, Attr, 1))
+    return;
+
+  if (checkAttrMutualExclusion<IntelFPGARegisterAttr>(S, D, Attr))
+    return;
+
+  SmallVector<Expr *, 8> Args;
+  for (unsigned I = 0; I < Attr.getNumArgs(); ++I) {
```
[the remark this answers is not in the document]
Comment (vmaksimo) [reply clang/lib/Sema/SemaDeclAttr.cpp]: There is only private method `getArgsBuffer()`. It seems that range loop is not possible here

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 4]
```diff
@@ -5209,6 +5220,92 @@ static void handleIntelFPGAMergeAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
...
+      }
+    }
+  } else {
+    llvm::APInt Num(32, (unsigned)(1 << Args.size()));
+    Expr *NBE =
+        IntegerLiteral::Create(Context, Num, Context.IntTy, SourceLocation());
+    D->addAttr(IntelFPGANumBanksAttr::CreateImplicit(Context, NBE));
+  }
+
+  if (!D->hasAttr<IntelFPGAMemoryAttr>())
```
Comment (MrSidims) [inline clang/lib/Sema/SemaDeclAttr.cpp]: Why is it needed?

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 5]
```diff
@@ -5209,6 +5220,92 @@ static void handleIntelFPGAMergeAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
...
+      if (Args.size() != Value.ceilLogBase2()) {
+        Diag(TmpAttr.getLoc(), diag::err_bankbits_numbanks_conflicting);
+        return;
+      }
+    }
+  } else {
+    llvm::APInt Num(32, (unsigned)(1 << Args.size()));
+    Expr *NBE =
+        IntegerLiteral::Create(Context, Num, Context.IntTy, SourceLocation());
+    D->addAttr(IntelFPGANumBanksAttr::CreateImplicit(Context, NBE));
```
Comment (MrSidims) [inline clang/lib/Sema/SemaDeclAttr.cpp]: May be better add it via call of AddOneConstantPowerTwoValueAttr() function?

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 6]
```diff
@@ -5209,6 +5220,92 @@ static void handleIntelFPGAMergeAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
                  IntelFPGAMergeAttr(S.Context, AL, Results[0], Results[1]));
 }
 
+/// Handle the bank_bits attribute.
+/// This attribute accepts a list of values greater than zero.
+/// This is incompatible with the register attribute.
+/// The numbanks and bank_bits attributes are related. If numbanks exists
+/// when handling bank_bits they are checked for consistency. If numbanks
+/// hasn't been added yet an implicit one is added with the correct value.
```
Comment (MrSidims) [inline clang/lib/Sema/SemaDeclAttr.cpp]: Why is that needed? I'm probably just missing something in all this memory attributes stuff...

**Comment by MrSidims** [comment]
Overall LGTM. But I'm starting to be more and more confused with all these memory attribute magic and implicit code insertion.

File: clang/test/CodeGenSYCL/intel-fpga-local.cpp [hunk 7]
```diff
@@ -18,12 +18,15 @@
 //CHECK: [[ANN16:@.str.[0-9]*]] = {{.*}}foobar
 //CHECK: [[ANN17:@.str.[0-9]*]] = {{.*}}{memory:MLAB}{sizeinfo:4,500}
 //CHECK: [[ANN18:@.str.[0-9]*]] = {{.*}}{memory:BLOCK_RAM}{sizeinfo:4,10,2}
+//CHECK: [[ANN19:@.str.[0-9]*]] = {{.*}}{memory:DEFAULT}{sizeinfo:4}{numbanks:4}{bank_bits:4,5}
```
Comment (MrSidims) [inline clang/test/CodeGenSYCL/intel-fpga-local.cpp]: Github is not allowing me to add comments to not modified strings for some reasons. Could you please add a compilation string for sycl-host? Probably I missed it one of previous reviews, sorry for that. I want to ensure, that in the correct cases we emit no diagnostics during host compilation. It can be done out of this change scope in a separate patch.

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 6 above)
Comment (vmaksimo) [reply clang/lib/Sema/SemaDeclAttr.cpp]: Because bank_bits and numbanks are related attributes - by using the bank_bits attribute, you can separate memory data into multiple banks while specifying which address bits to use to select the bank.

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 4 above)
Comment (vmaksimo) [reply clang/lib/Sema/SemaDeclAttr.cpp]: Because bank_bits attribute implies the memory attribute

File: clang/test/CodeGenSYCL/intel-fpga-local.cpp (hunk 7 above)
Comment (vmaksimo) [reply clang/test/CodeGenSYCL/intel-fpga-local.cpp]: Agree, it's needed to be improved. I will add this case in a separate PR

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 5 above)
Comment (vmaksimo) [reply clang/lib/Sema/SemaDeclAttr.cpp]: I am not sure about this approach. `AddOneConstantPowerTwoValueAttr()` function is in Sema class, while all the auxiliary attributes are added with `D->addAttr()` - code consistency will be broken . Also, looks like Sema function requires more code to be executed.

Comment (MrSidims) [reply clang/lib/Sema/SemaDeclAttr.cpp]: I see, thanks.

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 8]
```diff
@@ -5209,6 +5220,92 @@ static void handleIntelFPGAMergeAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
...
+/// If the user later adds a numbanks attribute the implicit one is removed.
+/// The values must be consecutive values (i.e. 3,4,5 or 2,1).
+static void handleIntelFPGABankBitsAttr(Sema &S, Decl *D,
+                                        const ParsedAttr &Attr) {
+  checkForDuplicateAttribute<IntelFPGABankBitsAttr>(S, D, Attr);
```
Comment (mlychkov) [inline clang/lib/Sema/SemaDeclAttr.cpp]: Add D decl validity check:
`
  if (D->isInvalidDecl())
    return;
`

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 9]
```diff
@@ -5209,6 +5220,92 @@ static void handleIntelFPGAMergeAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
...
+    }
+    Args.push_back(E);
+    Values.push_back(Value.getExtValue());
+  }
+
+  // Check that the list is consecutive.
+  if (!ListIsValueDep && Values.size() > 1) {
+    bool ListIsAscending = Values[0] < Values[1];
+    for (int I = 0, E = Values.size() - 1; I < E; ++I) {
+      if (Values[I + 1] != Values[I] + (ListIsAscending ? 1 : -1)) {
```
Comment (MrSidims) [inline clang/lib/Sema/SemaDeclAttr.cpp]: Not sure that it's mathematically correct requirement for consecutiveness (I can imagine +const; -const order), but I think there are hardware reasons for that to be +-1. So LGTM.

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 10]
```diff
@@ -5209,6 +5220,92 @@ static void handleIntelFPGAMergeAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
...
+        Value = Result.Val.getInt();
+      E = ICE.get();
+    }
+    Args.push_back(E);
+    Values.push_back(Value.getExtValue());
+  }
+
+  // Check that the list is consecutive.
+  if (!ListIsValueDep && Values.size() > 1) {
+    bool ListIsAscending = Values[0] < Values[1];
```
Comment (mlychkov) [inline clang/lib/Sema/SemaDeclAttr.cpp]: Maybe, better to rewrite in the following way:
`int64_t Step = Values[0] < Values[1] ? 1 : -1;
for (int I = 0, E = Values.size() - 1; I < E; ++I) {
  if (Values[I + 1] - Values[I] != Step) {`

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 8 above)
Comment (vmaksimo) [reply clang/lib/Sema/SemaDeclAttr.cpp]: Are you sure we need this check? It seems that other attributes handling has no such check. I think that it is handled higher in the call stack.

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 10 above)
Comment (vmaksimo) [reply clang/lib/Sema/SemaDeclAttr.cpp]: I like this approach, it will be even more useful in case of not only 1 and -1 steps available, but it requires int/int64 variable instead of bool. Let's leave it as it has already done

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 8 above)
Comment (MrSidims) [reply clang/lib/Sema/SemaDeclAttr.cpp]: It depends on a contract we have with clang code:
1. We trust function inputs;
2. We don't trust function inputs.

Comment (MrSidims) [reply clang/lib/Sema/SemaDeclAttr.cpp]: I don't have here a strong evidences to push anyone to follow my own preferences, but as far as I'm a pedantic person - I'd prefer to add this check at least in a code what I write.

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 10 above)
Comment (mlychkov) [reply clang/lib/Sema/SemaDeclAttr.cpp]: Ok.

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 8 above)
Comment (mlychkov) [reply clang/lib/Sema/SemaDeclAttr.cpp]: Quickly looking into the code I did not find such checks.

Comment (mlychkov) [reply clang/lib/Sema/SemaDeclAttr.cpp]: I mean declaration invalidation checks before attribute handling earlier in the code.
